File: ServerUpdate.py
__author__ = "luowen"
"""
Description: Automatic update ishadowsocks's server address and password.
Author: awimj <bigpao.luo@gmail.com>
Time: 2016-02-29
"""
import re, json, os, time, logging, urllib, urllib.request as request

HOSTADDR = "http://www.ishadowsocks.com/"
ISHADOWSCOKS_PROCESS_NAME = "Shadowsocks.exe"
ISHADOWSCOKS_PROCESS_PATH = "D:/Services/shadowsocks/Shadowsocks.exe"
SERVER_TYPE = "C"  # "A" => USA , "B" => HongKong, "C" => Japan


def updateServer(proxyServerType = "C"):
    """
        first: fetch server address and password from ishadowsocks website
        second: fetch success then merger ishadowsocks client configure file
    """
    serverInfo = fetchIshadowsocks(proxyServerType)
    if serverInfo and serverInfo.get("server") and serverInfo.get("password") and serverInfo.get("port"):
        # if address or password or port invalid don't merge
        if mergeLocalConfig(serverInfo):
            killProcess()
            time.sleep(2)
            startProcess()
    else:
        logMsg("'{0}' server address or password or port invalid.".format(proxyServerType))
    exit(0)

def fetchIshadowsocks(serverType):
    "fetch server address and password from ishadowsocks websiet"
    serverInfo = {}
    breakPointIndex = 0
    try:
        with request.urlopen(HOSTADDR) as urlHandle:
            for line in urlHandle:
                lineString = line.decode("utf-8");
                reObjMatchUrl = re.search("<h4>{0}服务器地址:(.*)</h4>".format(serverType), lineString)
                if reObjMatchUrl:
                    serverInfo["server"] = reObjMatchUrl.group(1)

                reObjMatchPwd = re.search("<h4>{0}密码:(.*)</h4>".format(serverType), lineString)
                if reObjMatchPwd:
                    serverInfo["password"] = reObjMatchPwd.group(1)
                reObjMatchPort = re.search("<h4>端口:(\d+)</h4>".format(serverType), lineString)
                if reObjMatchPort:
                    breakPointIndex += 1
                    if "{0}-{1}".format(serverType, breakPointIndex) == "A-1":
                        serverInfo["port"] = reObjMatchPort.group(1)
                    elif "{0}-{1}".format(serverType, breakPointIndex) == "B-2":
                        serverInfo["port"] = reObjMatchPort.group(1)
                    elif "{0}-{1}".format(serverType, breakPointIndex) == "C-3":
                        serverInfo["port"] = reObjMatchPort.group(1)
                    else:
                        continue
        return serverInfo
    except urllib.error.URLError as error:
        logMsg(error)

def mergeLocalConfig(serverInfo):
    " merge ishadowsocks website server information to local ishadowsocks client configure file"
    # TODO get local configure file and merge it
    guiConfigFileHandle = open("gui-config.json", encoding="utf-8")
    guiConfigJsonObj = json.load(guiConfigFileHandle)

    guiOldValue = guiConfigJsonObj['configs'][0]
    # assign new server information
    guiConfigJsonObj['configs'][0]["server"] = serverInfo['server']
    guiConfigJsonObj['configs'][0]["password"] = serverInfo['password']
    guiConfigJsonObj['configs'][0]["server_port"] = serverInfo['port']

    guiNewConfigFileHandle = open("gui-config.json", encoding="utf-8", mode="w")
    json.dump(guiConfigJsonObj, guiNewConfigFileHandle)
    logMsg("replace old server value [{0}] to new value [{1}].".format(str(guiOldValue), str(serverInfo)))
    return True

def logMsg(msg):
    "log process information to application.log"
    logging.basicConfig(filename="application.log", level=logging.INFO, format="[%(asctime)s]: %(message)s")
    logging.info(msg)

def killProcess():
    "update server information need restart, so first kill it."
    try:
        os.system("taskkill /F /IM {0} /T".format(ISHADOWSCOKS_PROCESS_NAME))
    except  OSError as error:
        logMsg(error)

def startProcess():
    "start shadowsocks process"
    try:
        os.system("start {0} &".format(ISHADOWSCOKS_PROCESS_PATH))
    except OSError as error:
        logMsg(error)


if __name__ == "__main__":
    updateServer(SERVER_TYPE)

File: ServerUpdate.php
<?php
/**
 * FileName: ServerUpdate.php
 * Description: fetch http://ishadowsocks.com last password
 * Author: Bigpao
 * Email: bigpao.luo@gmail.com
 * HomePage: 
 * Version: 0.0.1
 * LastChange: 2016-03-08 09:31:15
 * History:
 */

$websiteUrl = 'http://www.ishadowsocks.com/'; // 网站地址, 一般不用修改
$typeServer = 'B'; // ['A', 'B', 'C'] 选用服务器地址 A表示美国, B表示香港, C表示日本
$appPath = "D:/Services/shadowsocks/Shadowsocks.exe"; // 安装目录

class ServerUpdate
{
    protected $websiteUrl = 'http://www.ishadowsocks.com/';
    protected $typeServer = 'A';
    protected $appPath = "D:/Services/shadowsocks/Shadowsocks.exe";
    protected $serverInfo = array();

    const CONFIG_FILE = "gui-config.json";
    const KILL_APP = "taskkill /F /IM Shadowsocks.exe /T";
    const START_APP = "start %s &";

    public function __construct($websiteUrl, $typeServer, $appPath)
    {
        $this -> websiteUrl = $websiteUrl;
        $this -> typeServer = $typeServer;
        $this -> appPath = $appPath;
    }

    public function run()
    {
        try
        {
            $serverInfo = $this -> getServerInfo();
            $newServerInfo = $this -> mergeConfig($serverInfo);
            $this -> dumpConfig($newServerInfo);
            $this -> killApp();
            sleep(3);
            $this -> startApp();
        }
        catch(Exception $e)
        {
            $this -> log($e -> getMessage());
        }
    }

    public function getServerInfo()
    {
        if(!empty($this -> serverInfo)) return $this -> serverInfo;

        $serverInfo = array();
        $contents = file_get_contents($this -> websiteUrl);

        $serverPattern = "/". $this -> typeServer . "服务器地址:(.*)\</";
        $serverMatches = array();
        $result = preg_match($serverPattern, $contents, $serverMatches);
        if($result && isset($serverMatches[1])) // assign server address
            $serverInfo['server'] = $serverMatches[1];

        $pwdPattern = "/". $this -> typeServer . "密码:(.*)\</";
        $pwdMatches = array();
        $result = preg_match($pwdPattern, $contents, $pwdMatches);
        if($result && isset($pwdMatches[1]))
            $serverInfo['password'] = $pwdMatches[1];

        $portPattern = "/端口:(\d+)/";
        $portMatches = array();
        $result = preg_match_all($portPattern, $contents, $portMatches);

        if($result && isset($portMatches[1]))
        {
            switch($this -> typeServer)
            {
            case 'A':
                $serverInfo['server_port'] = $portMatches[1][0];
                break;
            case 'B':
                $serverInfo['server_port'] = $portMatches[1][1];
                break;
            case 'C':
                $serverInfo['server_port'] = $portMatches[1][2];
                break;
            default:
                break;
            }
        }

        return $this -> serverInfo = $serverInfo;
    }

    protected function mergeConfig($newServerInfo)
    {
        if(empty($newServerInfo))
        {
            $this -> log($message = 'Get server infomation from shadowsocks.com error.');
            throw new Exception($message);
        }
        $oldServerInfo = $this -> getConfigure();
        $oldServerInfo['configs'][0] = array_merge($oldServerInfo['configs'][0], $newServerInfo);
        return $oldServerInfo;
    }

    protected function getConfigure()
    {
        $softConfigString = file_get_contents(self::CONFIG_FILE);
        $softConfigJson = json_decode($softConfigString, true);
        if(is_array($softConfigJson) && isset($softConfigJson['configs']) && isset($softConfigJson['configs'][0]))
            return $softConfigJson;
        $this -> log($message = 'Application config file[gui-config.json] bad format. please check it.');
        throw new Exception($message);
    }

    protected function dumpConfig($config)
    {
        file_put_contents(self::CONFIG_FILE, json_encode($config, JSON_UNESCAPED_UNICODE));
    }

    protected function killApp()
    {
        echo exec(self::KILL_APP);
    }

    protected function startApp()
    {
        $startCommand = sprintf(self::START_APP, $this -> appPath);
        echo exec($startCommand);
    }

    protected function log($msg)
    {
        $formatMessage = '['. date('Y-m-d H:i:s', time()).']: ' . $msg . "\n";
        $fileHandle = file_put_contents('Application.log', $formatMessage, FILE_APPEND);
    }

}
(new ServerUpdate($websiteUrl, $typeServer, $appPath)) -> run();
